processor/src/main/java/com/linecorp/decaton/processor/runtime/ProcessorSubscription.java

```diff
@@ -103,14 +101,22 @@ private void updateState(SubscriptionStateListener.State newState) {
                      .collect(Collectors.toSet());
     }
 
-    private void commitCompletedOffsets(Consumer<?, ?> consumer) {
+    private void commitCompletedOffsets(Consumer<?, ?> consumer, boolean sync) {
         Map<TopicPartition, OffsetAndMetadata> commitOffsets = contexts.commitOffsets();
         if (commitOffsets.isEmpty()) {
-            logger.debug("No new offsets to commit, skipping commit");
+            log.debug("No new offsets to commit, skipping commit");
             return;
         }
-        logger.debug("Committing offsets: {}", commitOffsets);
-        consumer.commitSync(commitOffsets);
+        log.debug("Committing offsets {}: {}", sync ? "SYNC" : "ASYNC", commitOffsets);
+        if (sync) {
```

Comment:
why do we keep the sync way? (does it have any merit?)

Reply:
Yes. Currently we call offset commit at following timings and we need below 2 to be synchronous to avoid unnecessary duplication (re-consume) of records because for those cases basically it's the last chance to checkpoint the completed offsets.
* Every 1 second in consumer.poll() => queue loop
* At group rebalance (involves partition reassignment)
* At shutdown

Reply:
I see. I didn't try by myself so it's pure speculation but as you postulated for Shutdown, I'd have expected that rebalancing would wait for pending commit requests to be completed before actually proceeding.
Anyways, both cases are pretty rare so keeping sync will not affect performance. It's just about cleaning up.